{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "Quins-Aptamer-ssDNA-protein-alignment-prediction.ipynb",
      "provenance": [],
      "collapsed_sections": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "source": [
        "The purpose of my project is the computational generation of ssDNA aptamers in respect to a given peptide via Watson-Crick pseudo-pairing probability scores.\n",
        "\n",
        "\"A total of 446 crystal structures of nucleotide–protein complexes provide 18 types of direct pseudo pairs and five types of water-mediated pseudo pairs between nucleotide bases and amino acid side-chains. Compared with the previously observed pseudo pairs in DNA/RNA–protein complexes (26), eight direct and five water-mediated new types of pseudo pairs are clustered in our data set of nucleotide–protein complexes. In addition, several pseudo pairs between bases and the peptide backbone are observed in this study.\", (Jiro Kondo, 2011). https://pubmed.ncbi.nlm.nih.gov/21737431/\n",
        "\n",
        "This program has three purposes -- predict the optimal aptamer for a given peptide, predict pseudoparing locations, give pseudopairing bonding strenth approximation. The pseudoparing scores of the primer sequencences shared by all aptamer candidates will be used to establish a minimum pseudoparing score under which an aptamer candidate will be discarded. \n",
        "\n"
      ],
      "metadata": {
        "id": "PdSjUXVtfX90"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "import numpy as np\n",
        "import random\n",
        "\n"
      ],
      "metadata": {
        "id": "geIcd54DXHgK"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "UBOHoCK9ejLN"
      },
      "outputs": [],
      "source": [
        "class interaction_approximator():\n",
        "\n",
        "    def __init__(self):\n",
        "        \n",
        "        #holds the possible relationship of each nucleotide to a given amino \n",
        "        #list of attribute tuples as responde to bases A,G,T\n",
        "        #each base alphabet member gives aspects of its prospective pseudobond partners and their scores relative to itself. \n",
        "        #tuples arranged by (specific amino, bond frequency, number of hydrogen bonds) ordered from most probable to least\n",
        "        #frequency score outweighs number of h bonds in signficance of ordering predispositions. \n",
        "        #results from the study concerning C -> amino bonding predispositions were inconclusive, no test yeilding both a probability score and hydrogen bond count. \n",
        "\n",
        "        #each value pertaining to a key holds a tuple of two tuples --\n",
        "        #Watson-Crick pseudo-pairs between nucleotides and amino acid side chains -- first tuple\n",
        "        #Watson Crick pseudopairing between nucleotide and peptide backbone -- second tuple\n",
        "\n",
        "\n",
        "        self.base_predisposition_dict = {'A':{'N':((3.8, 4)), 'D':(( 2.6, 10)),'Q':((0.8, 4)), 'S':((0.4, 6)}, 'G':{'D':(34.1,2),'Q':(2.4,7)}, 'T':{'N':(8.3,2),'Q':(16.3,1)}}\n",
        "        self.amino_predisposition_dict = {'Q':{'G':(2.4, 7), 'A':(0.8, 4), 'T':(16.3, 1)}, 'D':{'G':(34.1, 2),'A':(2.6,10)}, 'N':{'T':(8.3,2), 'A':(3.8,4)}, 'S':{'A':(0.4,6)}}\n",
        "\n",
        "\n",
        "        #the value is in units of kcal/mol  \n",
        "        self.hydrogen_bond_strength = 0.0000000001     \n",
        "\n",
        "        #the primer sequences must be scored to aid in reducing false positives and obfuscating results. \n",
        "        #will be used to develop a minimum score to test sequences against for study validity \n",
        "        self.null_pseudopair_strength_and_peptide_fragment = (0, '')\n",
        "        self.p5 = \"TCGTCGGCAGCGTCAGATGTGTATAAGAGACAG\"\n",
        "        self.p7 = \"CTGTCTCTTATACACATCTCCGAGCCCACGAGAC\"\n",
        "        self.null_model_aptamer_sequence = self.p5 + self.p7\n",
        "        self.null_pseudopair_bond_strength = 0\n",
        "        self.null_pseudopair_bond_probability = 0\n",
        "\n",
        "        #the number of nucleotides to be selected:\n",
        "        self.N = 50\n",
        "\n",
        "        #list of attributes that went into the development of the given aptamers\n",
        "        self.random_aptamer_attributes_list = []\n",
        "        \n",
        "        #GFP\n",
        "        self.target_peptide = \"MSKGEELFTG VVPILVELDG DVNGHKFSVS GEGEGDATYG KLTLKFICTT GKLPVPWPTL VTTFSYGVQC FSRYPDHMKQ HDFFKSAMPE GYVQERTIFF KDDGNYKTRA EVKFEGDTLV NRIELKGIDF KEDGNILGHK LEYNYNSHNV YIMADKQKNG IKVNFKIRHN IEDGSVQLAD HYQQNTPIGD GPVLLPDNHY LSTQSALSKD PNEKRDHMVL LEFVTAAGIT HGMDELYK\"\n",
        "        self.peptide_fragment_list = []\n",
        "\n",
        "        self.aptamer_candidate_pool_list = []\n",
        "        #aminos which have adequate pseudopairing statistics \n",
        "\n",
        "        #set to false to alter sort contingent variable\n",
        "        self.sort_by_bond_strength = True\n",
        "\n",
        "        self.programatc_iterations = 10\n",
        "\n",
        "    def parse_and_score(self):\n",
        "        \"\"\"\n",
        "        The purpose of this function is to develop stability scores for each k and nmer.\n",
        "        These scores represent an expectation of molecular stability, as extrapolated \n",
        "        from the sited texts empirical values and bias results.\n",
        "\n",
        "        and then augmenting the scores of the kmers in regards to relationship of \n",
        "        the ssDNA kmer to the protein nmer.   \n",
        "        \n",
        "        Input: dictionary of pseudoparing predispositions from the amino perspective\n",
        "        Output:\n",
        "        \n",
        "           \"\"\"\n",
        "        \n",
        "        target_peptide_stripped = self.target_peptide.replace(\" \", '')\n",
        "\n",
        "        self.target_peptide = target_peptide_stripped\n",
        "        #parse the peptide, treating it as the emperical path -- the aptamer is unknown and to be established! \n",
        "        \n",
        "        #develop list of all possible fragments of peptide which will be used to develop aptamer candidates for comparison  \n",
        "        for index in range(0, len(self.target_peptide)- self.N):\n",
        "            self.peptide_fragment_list.append(self.target_peptide[index:index+self.N])\n",
        "\n",
        "        for peptide_fragment in self.peptide_fragment_list:\n",
        "            #the aminos from the peptide fragments will be used to develop an aptamer from weighted random\n",
        "            self.random_aptamer_attributes_list = [] \n",
        "            for amino in peptide_fragment:\n",
        "                #check if the amino is in the qualtified amino alphabet \n",
        "                if amino in self.amino_predisposition_dict.keys():\n",
        "\n",
        "                    possible_base_list = self.amino_predisposition_dict[amino].items()\n",
        "        \n",
        "                    possible_base_list = list(possible_base_list)\n",
        "                #pull weights from list of possible bases to be used in the random index generator \n",
        "                    base_probability_weight_tup = tuple([i[1][0] for i in possible_base_list])\n",
        "\n",
        "                    #randomly select index of the amino accociated bases by weights -- use that index to access the selected tuple \n",
        "                    selected_base = random.choices(possible_base_list, weights= base_probability_weight_tup, k=1)[0]\n",
        "                    self.random_aptamer_attributes_list.append((selected_base[0], self.amino_predisposition_dict[amino][selected_base[0]]))\n",
        "\n",
        "                else:\n",
        "                    #if there is not sufficient pseudopairing information for the current amino, N results\n",
        "                    self.random_aptamer_attributes_list.append(('N', (0.0, 0)))\n",
        "            \n",
        "            #at the end of each run, the current prospective aptamer will be compared to the null model\n",
        "            #if the bond strength is greater than the null model bond strength, the current aptamer\n",
        "            #enters he final aptamer candidate pool \n",
        "            self.pseudopair_bondstrength_threshold_checker()\n",
        "     \n",
        "        return \n",
        "\n",
        "    def pseudopair_bondstrength_threshold_checker(self):\n",
        "        \"\"\"Will add to aptamer candidate list of the current aptamer meets the threshold value for bonding strength.\"\"\"\n",
        "    \n",
        "        aptamer_candidate_quantities_tuple = self.statistic_calculator()\n",
        "         #if the threshold value is exceeded by the candidates score, \n",
        "         #append the aptamer candidate attribute to candidate pool list\n",
        "        if aptamer_candidate_quantities_tuple[2] > self.null_pseudopair_bond_probability:\n",
        "            self.aptamer_candidate_pool_list.append(aptamer_candidate_quantities_tuple)\n",
        "        return  \n",
        "\n",
        "\n",
        "    def statistic_calculator(self):\n",
        "        \"\"\"Counts the number of H bonds present in the prospective \n",
        "        pseudopairing regions shared by the ssDNA and peptide \n",
        "        sequence's aminos.\n",
        "        \n",
        "        Input: access to data accociated with nucleotide - amino interactions\n",
        "        Output: aptamer candidate list\n",
        "        \n",
        "         \"\"\"\n",
        "\n",
        "        #this will record the number of hydrogen bonds in the pseudopair\n",
        "        #this will record the prediction score\n",
        "\n",
        "        aptamer_candidate_sequence = ''\n",
        "\n",
        "        hydrogen_bond_sum = sum([i[1][1] for i in self.random_aptamer_attributes_list])\n",
        "        occurance_prediction_sum = sum([i[1][0] for i in self.random_aptamer_attributes_list])\n",
        "        pseudopair_bond_strength = hydrogen_bond_sum * self.hydrogen_bond_strength\n",
        "        \n",
        "        for base in [i[0] for i in self.random_aptamer_attributes_list]: \n",
        "            aptamer_candidate_sequence += base\n",
        "\n",
        "        return (aptamer_candidate_sequence, hydrogen_bond_sum, occurance_prediction_sum, pseudopair_bond_strength) \n",
        "\n",
        "\n",
        "    def null_model_aptamer_value_calculator(self):\n",
        "        \"\"\"Bonds occuring only between the primer sequences and the peptide will have the resulting score.\n",
        "            This score will be used as a null model to discard aptamer candidates. \n",
        "\n",
        "            This score is predicted from the prospective of the null model aptamer sequence, p5+p7, as mapped onto the peptides fragments. \n",
        "\n",
        "        Input: null model aptamer sequence -- PCR primers\n",
        "        Output: null model Watson-Crick pseudo-pairing probablity sum value and bonding strength\n",
        "        \"\"\"\n",
        "        target_peptide_stripped = self.target_peptide.replace(\" \", '')\n",
        "\n",
        "        null_peptide = target_peptide_stripped\n",
        "     \n",
        "        null_peptide_fragment_list = []\n",
        "        for index in range(0, len(null_peptide)- len(self.null_model_aptamer_sequence)):\n",
        "            null_peptide_fragment_list.append(null_peptide[index:index+len(self.null_model_aptamer_sequence)])\n",
        "\n",
        "\n",
        "        aptamer_score_list = []\n",
        "\n",
        "        #relationships between null model and fragments are 1-1, as both base and amino sequences are known\n",
        "        for peptide_fragment in null_peptide_fragment_list:\n",
        "            null_aptamer_attribute_list = []\n",
        "            for index in range(len(peptide_fragment)):\n",
        "                \n",
        "                #condition one -- the amino must be in the amino alphabet and the base must be in the base alphabet \n",
        "                if peptide_fragment[index] in self.amino_predisposition_dict.keys() and self.null_model_aptamer_sequence[index] in self.base_predisposition_dict.keys():  \n",
        "                    \n",
        "                    #condition 2 -- cross membership deterministic 1-1 relationship\n",
        "                    if peptide_fragment[index] in self.base_predisposition_dict[self.null_model_aptamer_sequence[index]].keys() and self.null_model_aptamer_sequence[index] in self.amino_predisposition_dict[peptide_fragment[index]].keys():\n",
        "                    \n",
        "                        null_aptamer_attribute_list.append(self.base_predisposition_dict[self.null_model_aptamer_sequence[index]][peptide_fragment[index]])\n",
        "\n",
        "            aptamer_score_list.append((sum([i[1] for i in null_aptamer_attribute_list]), peptide_fragment, sum([i[0] for i in null_aptamer_attribute_list])))\n",
        "\n",
        "        self.null_pseudopair_strength_and_peptide_fragment = max(aptamer_score_list)\n",
        "        \n",
        "        self.null_pseudopair_bond_strength = round(self.null_pseudopair_strength_and_peptide_fragment[0] * self.hydrogen_bond_strength, 12)\n",
        "        self.null_pseudopair_bond_probability = round(self.null_pseudopair_strength_and_peptide_fragment[2],3)\n",
        "        print(\"calculated null model values:\")\n",
        "        print(\"null occurance probability sum:\", self.null_pseudopair_bond_probability)\n",
        "        print(\"null bond strength:\", self.null_pseudopair_bond_strength, \"kcal/mol\")\n",
        "    \n",
        "    def final_candidate_display(self):\n",
        "        \"\"\"\n",
        "        This function develops the final output which is sorted by either bond strength at default. \n",
        "        \"\"\"\n",
        "\n",
        "        print(\" \")\n",
        "        #sorting by bond strength if option activated\n",
        "        if self.sort_by_bond_strength:\n",
        "            self.aptamer_candidate_pool_list.sort(key = lambda x: x[3], reverse=True) \n",
        "        else:\n",
        "            self.aptamer_candidate_pool_list.sort(key = lambda x: x[2], reverse=True)\n",
        "        print(\"Aptamer Candidate Pool:\")\n",
        "        print(\"rank, ssDNA aptamer sequence, probability sum, number of H-bonds, pseudopair hydrogen bond strength kcal/mol\")\n",
        "        i=0\n",
        "        for aptamer in self.aptamer_candidate_pool_list[0:100]:\n",
        "            i+=1\n",
        "            print(i, aptamer[0], round(aptamer[2],3), round(aptamer[1],3), round(aptamer[3],12))\n",
        "            \n",
        "        return \n",
        "\n",
        "\n",
        "    def driver(self):\n",
        "        \"\"\"The purpose of this function is to drive iterations of the program.\n",
        "        \"\"\"\n",
        "        print(\"programatic iterations: \", self.programatc_iterations)\n",
        "        print(\" \")\n",
        "        #find bonding strength and probability scores for primer sequences p5 and p7        \n",
        "        self.null_model_aptamer_value_calculator()\n",
        "\n",
        "        \n",
        "        for i in range(0, self.programatc_iterations):\n",
        "            self.parse_and_score()\n",
        "\n",
        "        self.final_candidate_display()\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "def main():\n",
        "\n",
        "    class_access = interaction_approximator()\n",
        "\n",
        "    class_access.driver()\n",
        "\n",
        "if __name__ == '__main__':\n",
        "    main()\n"
      ],
      "metadata": {
        "id": "RAPPa8HyfW-X",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "89591ce9-68e4-459a-8794-b6533b1b2993"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "programatic iterations:  10\n",
            " \n",
            "calculated null model values:\n",
            "null occurance probability sum: 53.3\n",
            "null bond strength: 7.6e-09 kcal/mol\n",
            " \n",
            "Aptamer Candidate Pool:\n",
            "rank, ssDNA aptamer sequence, probability sum, number of H-bonds, pseudopair hydrogen bond strength kcal/mol\n",
            "1 GNANNNNTNNNNNANNANANTNNANNGAANNNNGNNNNNNATNNNANTAN 109.3 86 8.6e-09\n",
            "2 NNNANNANANGNNANNTTTNNNNGNNNNNNGTNNNANTANNANGNTNNNA 191.7 80 8e-09\n",
            "3 NANAANANNNNNANTNANNNNTNNNNNTNNGNANGNNANNATTNNNNANN 118.6 79 7.9e-09\n",
            "4 NNNANNNNNANNGNANGNNGNNATTNNNNANNNNNNAANNNANTANNANG 164.6 77 7.7e-09\n",
            "5 NNTNNANANTNNANNTATNNNNGNNNNNNGANNNANTANNANGNTNNNAN 190.1 77 7.7e-09\n",
            "6 NNNTNNANANTNNGNNGTTNNNNGNNNNNNGTNNNANGANNANANTNNNG 213.8 76 7.6e-09\n",
            "7 NTNNNNNANNGNANTNNGNNGTANNNNGNNNNNNGTNNNANGANNANANA 206.0 74 7.4e-09\n",
            "8 TNNNNNANNGNANGNNANNGTTNNNNGNNNNNNGANNNANTANNANGNAN 206.0 74 7.4e-09\n",
            "9 NNNTNNGNANTNNANNTGTNNNNGNNNNNNGANNNANTANNANANANNNG 218.7 74 7.4e-09\n",
            "10 NANNGNANTNNANNTTANNNNGNNNNNNATNNNANGANNANGNTNNNGNN 218.7 74 7.4e-09\n",
            "11 NNGNANGNNGNNAGANNNNGNNNNNNGANNNANGANNANGNANNNGNNNN 225.6 73 7.3e-09\n",
            "12 ANNGNANGNNGNNTTTNNNNGNNNNNNATNNNANTANNANGNTNNNANNN 223.2 72 7.2e-09\n",
            "13 NNTNNNNNTNNANANTNNGNNTTTNNNNANNNNNNAANNNANTANNANGN 171.5 72 7.2e-09\n",
            "14 NNNANNNNNTNNANANTNNGNNTGANNNNGNNNNNNGANNNANGANNANG 197.7 72 7.2e-09\n",
            "15 NTNNGNANGNNGNNTTANNNNGNNNNNNGANNNANGANNANANTNNNGNN 236.3 72 7.2e-09\n",
            "16 NNANANTNNANNTTTNNNNGNNNNNNGANNNANGANNANGNANNNGNNNN 210.4 72 7.2e-09\n",
            "17 NNNNTNNGNANTNNGNNTGANNNNANNNNNNAANNNANTANNANGNTNNN 184.6 72 7.2e-09\n",
            "18 NANANANNGNNGTTNNNNGNNNNNNGANNNANGANNANGNTNNNGNNNNN 217.0 71 7.1e-09\n",
            "19 NNTNNGNANTNNANNTAANNNNGNNNNNNATNNNANTANNANGNANNNGN 217.1 71 7.1e-09\n",
            "20 NNNNTNNNNNTNNGNANGNNANNTTTNNNNANNNNNNGANNNANAANNAN 139.5 71 7.1e-09\n",
            "21 NNTNNGNANANNANNTTANNNNGNNNNNNAANNNANTANNANGNTNNNGN 217.1 71 7.1e-09\n",
            "22 TNNNNANNNNNANNANANTNNGNNTAANNNNANNNNNNGTNNNANTANNA 152.7 71 7.1e-09\n",
            "23 NNNNNTNNANANTNNANNTTTNNNNGNNNNNNGANNNANGANNANGNTNN 189.1 70 7e-09\n",
            "24 NNTNNGNANGNNGNNGTTNNNNGNNNNNNATNNNANTANNANGNANNNGN 240.8 70 7e-09\n",
            "25 NNTNNNNNANNGNANTNNGNNTGANNNNGNNNNNNGTNNNANGANNANAN 202.2 70 7e-09\n",
            "26 NGNANTNNGNNTGTNNNNGNNNNNNAANNNANTANNANANTNNNGNNNNN 214.9 70 7e-09\n",
            "27 TNNNNNTNNGNANTNNGNNTGTNNNNANNNNNNGANNNANGANNANGNTN 215.0 70 7e-09\n",
            "28 NNTNNGNANGNNGNNTTANNNNANNNNNNGANNNANAANNANGNTNNNGN 234.7 69 6.9e-09\n",
            "29 TNNNNNANGNANNNNANNNNNANNGNANTNNGNNGATNNNNANNNNNNGT 166.1 69 6.9e-09\n",
            "30 ANNGNANTNNANNATTNNNNGNNNNNNGTNNNANTANNANANTNNNGNNN 221.6 69 6.9e-09\n",
            "31 NNNNANNGNANTNNGNNTAANNNNGNNNNNNATNNNANGANNANGNANNN 196.1 69 6.9e-09\n",
            "32 NNGNANTNNGNNTATNNNNGNNNNNNAANNNANTANNANANANNNGNNNN 208.8 69 6.9e-09\n",
            "33 NNNTNNNNNANNGNANTNNGNNTATNNNNGNNNNNNAANNNANTANNANA 183.0 69 6.9e-09\n",
            "34 ANNNNNTNNGNANGNNANNATTNNNNGNNNNNNGTNNNANTANNANGNAN 208.9 69 6.9e-09\n",
            "35 NANTNNGNNGTTNNNNGNNNNNNGANNNANGANNANGNANNNANNNNNNN 193.9 68 6.8e-09\n",
            "36 NTNNGNANTNNGNNTGTNNNNGNNNNNNATNNNANGANNANGNTNNNGNN 245.3 68 6.8e-09\n",
            "37 NNANNGNANTNNGNNTTTNNNNANNNNNNGTNNNANTANNANGNANNNAN 232.6 68 6.8e-09\n",
            "38 NTNNNNNTNNGNANTNNGNNGGTNNNNGNNNNNNATNNNANTANNANGNT 219.5 68 6.8e-09\n",
            "39 NNTNNGNANTNNGNNTTANNNNGNNNNNNAANNNANTANNANGNTNNNAN 232.6 68 6.8e-09\n",
            "40 NNNNTNNGNANTNNGNNTGTNNNNANNNNNNGTNNNANGANNANGNANNN 206.7 68 6.8e-09\n",
            "41 NNANTANTNNNNNANTNANNNNANNNNNANNGNANGNNGNNTGANNNNGN 178.7 68 6.8e-09\n",
            "42 NNNNANNGNANTNNGNNGGTNNNNGNNNNNNGTNNNANTANNANANTNNN 206.7 68 6.8e-09\n",
            "43 GNANTNNGNNTGANNNNGNNNNNNATNNNANGANNANGNTNNNGNNNNNN 232.5 68 6.8e-09\n",
            "44 NANANGNNGNNTGANNNNGNNNNNNGTNNNANTANNANGNTNNNGNNNNN 232.5 68 6.8e-09\n",
            "45 NANTNNGNNTGTNNNNGNNNNNNAANNNANTANNANANTNNNGNNNNNNN 180.8 68 6.8e-09\n",
            "46 NNGNANGNNGNNTGTNNNNGNNNNNNGTNNNANTANNANANANNNGNNNN 232.5 68 6.8e-09\n",
            "47 NAANTNNNNNGNTNANNNNANNNNNANNGNANTNNANNGTANNNNANNNN 152.8 68 6.8e-09\n",
            "48 NNNNANNNNNTNNGNANTNNANNGTANNNNGNNNNNNGTNNNANGANNAN 168.1 68 6.8e-09\n",
            "49 TNNNNNNNNTNAANTNNNNNANTNTNNNNTNNNNNANNANANGNNGNNGT 126.6 68 6.8e-09\n",
            "50 NAANANNNNNANTNTNNNNTNNNNNANNANANTNNANNTTANNNNGNNNN 139.7 68 6.8e-09\n",
            "51 NANNANANGNNGNNTTANNNNGNNNNNNGANNNANTANNANGNTNNNGNN 245.7 68 6.8e-09\n",
            "52 NNGNANTNNGNNTGTNNNNGNNNNNNGANNNANGANNANANTNNNGNNNN 232.5 68 6.8e-09\n",
            "53 NNTNNGNANTNNGNNGTTNNNNANNNNNNGTNNNANAANNANGNANNNGN 239.2 67 6.7e-09\n",
            "54 ANGNTNNNNANNNNNANNGNANTNNGNNTGTNNNNGNNNNNNGANNNANA 206.0 66 6.6e-09\n",
            "55 NNANTNTNNNNTNNNNNTNNANANTNNGNNGTTNNNNGNNNNNNATNNNA 169.6 66 6.6e-09\n",
            "56 NNNTNNGNANTNNANNTTANNNNGNNNNNNGTNNNANTANNANGNTNNNA 237.1 66 6.6e-09\n",
            "57 NNTNNGNANGNNGNNTTTNNNNGNNNNNNGANNNANTANNANANANNNGN 250.2 66 6.6e-09\n",
            "58 NTNNNNNANNGNANGNNGNNTTTNNNNGNNNNNNGTNNNANTANNANANA 224.4 66 6.6e-09\n",
            "59 TNNNNNTNNGNANGNNGNNTTANNNNGNNNNNNAANNNANTANNANGNTN 224.4 66 6.6e-09\n",
            "60 TNNNNNTNNGNANTNNGNNGTANNNNANNNNNNGTNNNANTANNANGNAN 224.4 66 6.6e-09\n",
            "61 NNNANNGNANTNNANNGTTNNNNGNNNNNNGANNNANTANNANGNTNNNG 250.2 66 6.6e-09\n",
            "62 NNNANNGNANTNNGNNTGTNNNNGNNNNNNAANNNANTANNANGNTNNNG 250.2 66 6.6e-09\n",
            "63 NNTNNGNANGNNGNNGTANNNNGNNNNNNGANNNANTANNANGNANNNGN 263.3 66 6.6e-09\n",
            "64 NNGNANGNNGNNTTTNNNNGNNNNNNGTNNNANGANNANANTNNNGNNNN 237.0 66 6.6e-09\n",
            "65 NNNNTNNNNNANNGNANTNNGNNGGTNNNNGNNNNNNATNNNANTANNAN 172.6 66 6.6e-09\n",
            "66 NGNNGNNGTTNNNNANNNNNNGTNNNANTANNANGNTNNNANNNNNNNNN 171.0 66 6.6e-09\n",
            "67 NANNNNNTNNGNANTNNGNNATTNNNNGNNNNNNAANNNANAANNANGNT 207.3 66 6.6e-09\n",
            "68 NANNGNANGNNGNNTTANNNNGNNNNNNGTNNNANTANNANGNTNNNANN 250.2 66 6.6e-09\n",
            "69 TNNGNANTNNGNNTTANNNNGNNNNNNGANNNANGANNANANTNNNGNNN 250.2 66 6.6e-09\n",
            "70 GNANGNNGNNTGTNNNNGNNNNNNGANNNANGANNANGNTNNNGNNNNNN 250.1 66 6.6e-09\n",
            "71 NNTNNNNNTNNGNANTNNGNNTTANNNNGNNNNNNAANNNANTANNANAN 198.5 66 6.6e-09\n",
            "72 NNGNTNNNNNNNNTNAANANNNNNANTNANNNNANNNNNTNNANANGNNG 133.0 66 6.6e-09\n",
            "73 NNNNNTNNGNANTNNGNNGTTNNNNANNNNNNGTNNNANGANNANGNTNN 211.2 66 6.6e-09\n",
            "74 NTNNGNANTNNGNNTGANNNNGNNNNNNAANNNANTANNANGNTNNNGNN 250.2 66 6.6e-09\n",
            "75 NNANTANANNNNNGNGNTNNNNTNNNNNANNGNANTNNANNTGANNNNGN 183.2 66 6.6e-09\n",
            "76 NNNTNNNNNTNNGNANGNNANNATTNNNNGNNNNNNGANNNANTANNANG 205.1 65 6.5e-09\n",
            "77 NTNTNNNNANNNNNTNNGNANGNNANNTTTNNNNGNNNNNNATNNNANTA 179.2 65 6.5e-09\n",
            "78 NNANNGNANTNNGNNTTANNNNGNNNNNNAANNNANAANNANGNTNNNGN 244.1 65 6.5e-09\n",
            "79 NNGNNTTANNNNANNNNNNGANNNANTANNANANTNNNANNNNNNNNNNN 142.0 65 6.5e-09\n",
            "80 NNNNNTNNGNANTNNANNTTTNNNNGNNNNNNGTNNNANAANNANANTNN 192.0 65 6.5e-09\n",
            "81 NGNANNNNTNNNNNANNANANTNNGNNTTTNNNNGNNNNNNGTNNNANGA 192.3 65 6.5e-09\n",
            "82 NTNNGNANANNGNNGTTNNNNGNNNNNNGANNNANGANNANGNTNNNGNN 256.8 65 6.5e-09\n",
            "83 NNNANNNGNNANNNANNNGNGNNNANNNNNNNGNNNNNNNAGNTNNNNNN 90.1 65 6.5e-09\n",
            "84 NNNTNAANTNNNNNGNTNTNNNNTNNNNNTNNANANGNNGNNATTNNNNA 163.6 65 6.5e-09\n",
            "85 GNTNANNNNTNNNNNANNGNANGNNGNNTGANNNNANNNNNNGANNNANT 217.0 65 6.5e-09\n",
            "86 NNNNNTNNGNANGNNANNTTANNNNGNNNNNNGTNNNANAANNANGNTNN 205.1 65 6.5e-09\n",
            "87 NNNNNTNNANANTNNANNTATNNNNGNNNNNNGTNNNANTANNANGNTNN 192.0 65 6.5e-09\n",
            "88 NNNNANNNNNTNNGNANGNNGNNAGANNNNGNNNNNNGANNNANTANNAN 179.6 65 6.5e-09\n",
            "89 NNGNANANNGNNGTTNNNNGNNNNNNAANNNANTANNANGNTNNNGNNNN 230.9 65 6.5e-09\n",
            "90 NNANNGNANTNNGNNTTTNNNNGNNNNNNGTNNNANGANNANGNTNNNAN 254.7 64 6.4e-09\n",
            "91 NANTNNGNNTTTNNNNANNNNNNGANNNANTANNANANANNNGNNNNNNN 190.2 64 6.4e-09\n",
            "92 NNTNNGNANTNNGNNTGTNNNNGNNNNNNGANNNANTANNANGNTNNNAN 254.7 64 6.4e-09\n",
            "93 NTNNGNANTNNANNTTTNNNNGNNNNNNGTNNNANGANNANGNANNNGNN 254.7 64 6.4e-09\n",
            "94 GNANNNNNNNNTNAANTNNNNNGNGNTNNNNANNNNNANNGNANGNNANN 150.6 64 6.4e-09\n",
            "95 NGNNTGANNNNANNNNNNGTNNNANAANNANGNTNNNANNNNNNNNNNNN 148.6 64 6.4e-09\n",
            "96 TANTNNNNNGNTNANNNNTNNNNNANNANANTNNANNTGANNNNGNNNNN 161.8 64 6.4e-09\n",
            "97 ANNGNANTNNGNNTGTNNNNGNNNNNNGTNNNANTANNANANTNNNGNNN 254.7 64 6.4e-09\n",
            "98 NANANTNNGNNTTANNNNGNNNNNNGANNNANGANNANGNTNNNGNNNNN 241.9 64 6.4e-09\n",
            "99 NNNTNAANTNNNNNGNTNANNNNTNNNNNTNNGNANTNNANNTGTNNNNA 174.6 64 6.4e-09\n",
            "100 ANNGNANTNNANNGTTNNNNGNNNNNNGTNNNANTANNANGNTNNNGNNN 254.7 64 6.4e-09\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "References:\n",
        "\n",
        "\n",
        "\n",
        "https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3201857/\n"
      ],
      "metadata": {
        "id": "VUxPiUjDjd8F"
      }
    }
  ]
}